Segment <> Shopify source (#5847)
segmentio/segment-docs: src/connections/sources/catalog/libraries/website/shopify/index.md

> ---
> title: Shopify
> redirect_from:
> id: 
> ---

Reviewer: ```suggestion
---
title: Shopify
redirect_from:
id: 
hidden: true
---
```

---
title: Shopify
redirect_from:
id: 
hidden: true
---

> The Shopify Source enables you to send [Shopify Web Pixel API Standard Events](https://shopify.dev/docs/api/web-pixels-api/standard-events){:target="\_blank"} from your Shopify Store to Segment. Events sent to Segment are formatted to match the [Segment Ecommerce Spec](https://segment.com/docs/connections/spec/ecommerce/v2/) as closely as possible. 

Reviewer: ```suggestion
This Shopify Source lets you send [Shopify Web Pixel API Standard Events](https://shopify.dev/docs/api/web-pixels-api/standard-events){:target="_blank"} from your Shopify Store to Segment. Events sent to Segment are formatted to match the [Segment Ecommerce Spec](/docs/connections/spec/ecommerce/v2/). 
```

> This Source is a free to use [Shopify App Extension](https://shopify.dev/docs/apps/app-extensions){:target="\_blank"} which can be installed via your Shopify Store's Admin interface.   

Reviewer: ```suggestion
This Source is a free [Shopify App Extension](https://shopify.dev/docs/apps/app-extensions){:target="_blank"} which can be installed using your Shopify Store's Admin interface.  
```

#### Overview

> Once installed and enabled, Segment events be sent from the user's browser to your Segment 'Shopify' Source in real time. 

Reviewer: ```suggestion
Once installed and enabled, Segment collects events from the user's browser and sends them to your Segment 'Shopify' Source in real time. 
```

> - You can control which Segment `track()` events get collected as well as how user profile details are passed to Segment.  
> - A custom JavaScript snippet is loaded in the Shopify store. This script transforms Shopify Standard Events to Segment Ecommerce Spec events. Note that Segment's Analytics.js 2.0 library will **not** load on any Shopify store page. 
> - This is a Device Mode (client side) Integration. No data is sent server side from Shopify to Segment.  
> - No post purchase events are collected.
> - The Segment anonmousId is set to the value of the Shopify client ID. The Segment userId value is never populated. 
> - Email address and phone number details will be sent to Segment if they are collected by Shopify.   

Reviewer: ```suggestion
- You can control which Segment `track()` events get collected as well as how user profile details are passed to Segment.  
- A custom JavaScript snippet is loaded in the Shopify store. This script transforms Shopify Standard Events to Segment Ecommerce Spec events. Note that Segment's Analytics.js 2.0 library will **not** load on any Shopify store page. 
- You can control which Segment Track events get collected as well as how user profile details are passed to Segment.  
- A custom JavaScript snippet is loaded in the Shopify store. This script transforms Shopify Standard Events to Segment Ecommerce Spec events. Note that Segment's Analytics.js 2.0 library does **not** load on any Shopify store page. 
- This is a Device Mode (client side) Integration. No data is sent server-side from Shopify to Segment.  
- This source does not collect post-purchase events.
- The Segment anonymousId is set to the value of the Shopify client ID. The Segment userId value is never populated. 
- Email address and phone number details are sent to Segment if they are collected by Shopify.   
```

> If you require more advanced functionality or post purchase events consider using the [Shopify by Littledata](https://segment.com/docs/connections/sources/catalog/libraries/website/shopify-littledata/) Source. 'Shopify by Littledata' requires a paid subscription. 

Reviewer: ```suggestion
If you require more advanced functionality or want to collect post-purchase events, consider using the [Shopify by Littledata](/docs/connections/sources/catalog/libraries/website/shopify-littledata/) Source. Shopify by Littledata requires a paid subscription. 
```

## Getting Started

> 1. **Log in** to your Shopify Store account.

Reviewer: ```suggestion
1. Log in to your Shopify Store account.
```

> 2. Go to the [Shopify app store listing](https://apps.shopify.com/twilio-segment){:target="\_blank"} for **Twilio Segment**.

Reviewer: ```suggestion
2. Open the Shopify app store and select the [Twilio Segment listing](https://apps.shopify.com/twilio-segment){:target="_blank"}
```

3. Click **Add app** to begin the installation process.

> 4. Add the [**Segment write key**](/docs/connections/find-writekey/) for your Segment Source, and select the Region for your Segment Workspace. 

Reviewer: ```suggestion
4. Add the [Segment write key](/docs/connections/find-writekey/) for your Segment Source, and select the Region for your Segment Workspace. 
```

>    ![Add Source writeKey and select Workspace Region.](images/settings.png)

Reviewer: we typically avoid using screenshots in the docs - they add load time to the site, they're not accessible for folks that use assistive technologies, and they're hard to maintain. given that this is a fairly straightforward form to fill out, could we remove this screenshot? 
```suggestion
```

> 5. Use the checkboxes under 'Manage Events' to select which `track()` events to send to Segment. 

Reviewer: ```suggestion
5. Use the checkboxes under 'Manage Events' to select which Track events you'd like to send to Segment.  
```

6. Use the checkboxes under 'Manage user profile details' to select how user profile details should be sent to Segment. 

> 6. Click the 'Turn On' button to enable start sending analytics events from your Shopify Store to Segment. 

Reviewer: ```suggestion
6. Click the **Turn On** button to start sending analytics events from your Shopify Store to Segment. 
```

> ## identifiers

Reviewer: ```suggestion
## Identifiers
```

> - **anonymousId** - The Segment `anonymousId` field will be populated with the Shopify client ID. The Shopify client ID is a 'client-side ID of the customer, provided by Shopify'. It is not currently possible to configure the Segment anonymousId to be populated with any other value. 
> - **userId** - The Segment `userId` field will not be populated. It is not currently possible to configure the userId to be populated with any value from Shopify. 
> - **email and phone** - `traits.email` and `traits.phone` will be collected via `identify()` events (unless explicity disabled via the 'Manage user profile details' section in the Shopify App User Interface). 

Reviewer: ```suggestion
- **anonymousId** - The Segment `anonymousId` field are populated with the Shopify client ID. The Shopify client ID is a 'client-side ID of the customer, provided by Shopify'. You cannot configure the Segment anonymousId to be populated with any other value. 
- **userId** - The Segment `userId` field does not populate. You cannot configure the userId to be populated with any value from Shopify. 
- **email and phone** - `traits.email` and `traits.phone` are collected using Identify events (unless explicitly disabled in the 'Manage user profile details' section of the Shopify App User Interface). 
```

> ## track() and page() events sent to Segment

Reviewer: ```suggestion
## Track and Page events sent to Segment
```

> See the [Shopify Standard Events](https://shopify.dev/docs/api/web-pixels-api/standard-events){:target="\_blank"} documentation for more information regarding when events are triggered. 

Reviewer: ```suggestion
See the [Shopify Standard Events](https://shopify.dev/docs/api/web-pixels-api/standard-events){:target="_blank"} documentation for more information about when events are triggered. 
```

| Segment Event Name      | Shopify Standard Event Name         | Event trigger                                                       | 
| ----------------------- | ----------------------------------- | ------------------------------------------------------------------- |

> | `Page Viewed`           | `page_viewed`                       | A user has visited a page. Send as a Segment `page()` event         |

Reviewer: ```suggestion
| `Page Viewed`           | `page_viewed`                       | A user has visited a page. Sent as a Segment Page event          |
```

| `Products Searched`     | `search_submitted`                  | A user has performed a search on the Shopify storefront             |
| `Product List Viewed`   | `collection_viewed`                 | A user has visited a product collection index page                  |
| `Product Viewed`        | `product_viewed`                    | A user has visited a product details page                           |
| `Product Added`         | `product_added_to_cart`             | A user has added a product to their cart                            |
| `Product Removed`       | `product_removed_from_cart`         | A user has removed a product from their cart                        |
| `Cart Viewed`           | `cart_viewed`                       | A user has visited the cart page                                    |
| `Checkout Started`      | `checkout_started`                  | A user has started the checkout process                             |

> | `Shipping Info Entered` | `checkout_shipping_info_submitted`  | A user has choosen a shipping rate                                  |

Reviewer: ```suggestion
| `Shipping Info Entered` | `checkout_shipping_info_submitted`  | A user has chosen a shipping rate                                  |
```

| `Address Info Entered`  | `checkout_address_info_submitted`   | A user has submitted their mailing address                          |
| `Contact Info Entered`  | `checkout_contact_info_submitted`   | A user has submitted a checkout form                                |
| `Payment Info Entered`  | `payment_info_submitted`            | A user has submitted their payment information                      |
| `Order Completed`       | `checkout_completed`                | A user has completed a purchase                                     |

> You can control which `track()` and `page()` events to send to Segment via the 'Manage events' section in the Shopify App User Interface.

Reviewer: ```suggestion
You can control which Track and Page events to send to Segment in the 'Manage events' section of the Shopify App User Interface.
```

![Screenshot of the Manage events settings section.](images/manage_events.png)
      
## Page context data

> The fields below are automatically sent to Segment with every `track()`, `page()` and `idetify()` event. For `page()` events these fields are also included in the `properties` object.

Reviewer: ```suggestion
The fields below are automatically sent to Segment with every Track, Page and Identify event. For Page events, these fields are also included in the `properties` object.
```

| Field                                   | Description                                                                                    | Type          |
| --------------------------------------- | ---------------------------------------------------------------------------------------------- | ------------- |
| `context.page.url`                      | The entire URL of the page                                                                     | String        |
| `context.page.search`                   | a string containing a '?' followed by the parameters or "querystring" of the URL               | String        |
| `context.page.path`                     | a string containing an initial '/' followed by the path of the URL, excludes the query string  | String        |
| `context.page.title`                    | The title of the current page                                                                  | String        |
| `context.page.referrer`                 | URI of the page that linked to the current page                                                | String        |

## UTM data

> The fields below are automatically sent to Segment with every `track()`, `page()` and `idetify()` if present in the QueryString of the page url.

Reviewer: ```suggestion
The fields below are automatically sent to Segment with every Track, Page and Identify call, if present in the QueryString of the page url.
```

| Field                                   | Description                                                                                      | Type          |
| --------------------------------------- | ------------------------------------------------------------------------------------------------ | ------------- |
| `context.campaign.source`               | `utm_source` QueryString value                                                                   | String        |
| `context.campaign.medium`               | `utm_medium` QueryString value                                                                   | String        |
| `context.campaign.name`                 | `utm_campaign` QueryString value                                                                 | String        |
| `context.campaign.term`                 | `utm_term` QueryString value                                                                     | String        |
| `context.campaign.content`              | `utm_content` QueryString value                                                                  | String        |

> ## Page Viewed / page() event properties
> 
> The properties below are included with `page()` events. Note: If a property is missing it is due to Shopify not providing the value, or due to a `null` or empty value being provided.  

Reviewer: ```suggestion
## Page Viewed / Page event properties

The properties below are included with Page events. 

> info "Missing properties?"
> Your Segment events may be missing one of the following properties if Shopify provides `null` or an empty value for a property.

```

| Field                                   | Description                                                                                    | Type          |
| --------------------------------------- | ---------------------------------------------------------------------------------------------- | ------------- |
| `url`                                   | The entire URL of the page                                                                     | String        |
| `search`                                | a string containing a '?' followed by the parameters or "querystring" of the URL               | String        |
| `path`                                  | a string containing an initial '/' followed by the path of the URL, excludes the query string  | String        |
| `title`                                 | The title of the current page                                                                  | String        |
| `referrer`                              | URI of the page that linked to the current page                                                | String        |

> ## Product Viewed track() event properties
> 
> The properties below are included with `Product Viewed` `track()` events. Note: If a property is missing it is due to Shopify not providing the value, or due to a `null` or empty value being provided.  

Reviewer: ```suggestion
## Product Viewed Track event properties

The properties below are included with `Product Viewed` Track events. 

> info "Missing properties?"
> Your Segment events may be missing one of the following properties if Shopify provides `null` or an empty value for a property.

```

| `category`                              | Product type specified by the merchant                                                         | String        |
| `brand`                                 | The product’s vendor name                                                                      | String        |
| `url`                                   | URL to the product page                                                                        | String        |
| `image_url`                             | URL to the product image                                                                       | String        |
| `currency`                              | The three-letter code that represents the currency, for example 'USD'                          | String        |
| `price`                                 | The decimal money amount (price) for the product                                               | Number        |
| `sku`                                   | The SKU (stock keeping unit) associated with the variant                                       | String        |

> ## Product Added and Product Removed track() event properties
> 
> The properties below are included with `Product Added` and `Product Removed` `track()` events. Note: If a property is missing it is due to Shopify not providing the value, or due to a `null` or empty value being provided.  

Reviewer: ```suggestion
## Product Added and Product Removed Track event properties

The properties below are included with `Product Added` and `Product Removed` Track events. 

> info "Missing properties?"
> Your Segment events may be missing one of the following properties if Shopify provides `null` or an empty value for a property.

```

| `url`                                   | URL to the product page                                                                        | String        |
| `image_url`                             | URL to the product image                                                                       | String        |
| `currency`                              | The three-letter code that represents the currency, for example 'USD'                          | String        |
| `price`                                 | The decimal money amount (price) for the product                                               | Number        |
| `quantity`                              | The quantity of the merchandise that the customer intends to purchase                          | Integer       |
| `total_price`                           | The total decimal money amount (price) for the product                                         | Number        |
| `sku`                                   | The SKU (stock keeping unit) associated with the variant                                       | String        |

> ## Product List Viewed track() event properties
> 
> The properties below are included with `Product List Viewed` `track()` events. `products.$.` indicates an array named 'products'. Note: If a property is missing it is due to Shopify not providing the value, or due to a `null` or empty value being provided.  

Reviewer: ```suggestion
## Product List Viewed Track event properties

The properties below are included with `Product List Viewed` Track events. `products.$.` indicates an array named 'products'. 

> info "Missing properties?"
> Your Segment events may be missing one of the following properties if Shopify provides `null` or an empty value for a property.

```

| `products.$.category`                   | Product type specified by the merchant                                                         | String        |
| `products.$.brand`                      | The product’s vendor name                                                                      | String        |
| `products.$.url`                        | URL to the product page                                                                        | String        |
| `products.$.image_url`                  | URL to the product image                                                                       | String        |
| `products.$.currency`                   | The three-letter code that represents the currency, for example 'USD'                          | String        |
| `products.$.price`                      | The decimal money amount (price) for the product                                               | Number        |
| `products.$.sku`                        | The SKU (stock keeping unit) associated with the variant                                       | String        |

> ## Products Searched track() event properties
> 
> The properties below are included with `Products Searched` `track()` events. `products.$.` indicates an array named 'products'. Note: If a property is missing it is due to Shopify not providing the value, or due to a `null` or empty value being provided.  

Reviewer: ```suggestion
## Products Searched Track event properties

The properties below are included with `Products Searched` Track events. `products.$.` indicates an array named 'products'. 

> info "Missing properties?"
> Your Segment events may be missing one of the following properties if Shopify provides `null` or an empty value for a property.

```

| `products.$.brand`                      | The product’s vendor name                                                                      | String        |
| `products.$.url`                        | URL to the product page                                                                        | String        |
| `products.$.image_url`                  | URL to the product image                                                                       | String        |
| `products.$.currency`                   | The three-letter code that represents the currency, for example 'USD'                          | String        |
| `products.$.price`                      | The decimal money amount (price) for the product                                               | Number        |
| `products.$.sku`                        | The SKU (stock keeping unit) associated with the variant                                       | String        |
| `query`                                 | The search query that was executed                                                             | String        |

> ## Cart Viewed track() event properties
> 
> The properties below are included with `Cart Viewed` `track()` events. `products.$.` indicates an array named 'products'. Note: If a property is missing it is due to Shopify not providing the value, or due to a `null` or empty value being provided.  

Reviewer: ```suggestion
## Cart Viewed Track event properties

The properties below are included with `Cart Viewed` Track events. `products.$.` indicates an array named 'products'. 

> info "Missing properties?"
> Your Segment events may be missing one of the following properties if Shopify provides `null` or an empty value for a property.

```

| `products.$.image_url`                  | URL to the product image                                                                       | String        |
| `products.$.currency`                   | The three-letter code that represents the currency, for example 'USD'                          | String        |
| `products.$.price`                      | The decimal money amount (price) for the product                                               | Number        |
| `products.$.sku`                        | The SKU (stock keeping unit) associated with the variant                                       | String        |
| `cart_id`                               | A globally unique identifier for the cart                                                      | String        |
| `subtotal`                              | The price at checkout before duties, shipping, and taxes                                       | Number        |
| `currency`                              | The three-letter code that represents the currency, for example 'USD'                          | String        |

> ## Checkout and purchase track() event properties
> 
> The properties below are included with `Checkout Started`, `Address Info Entered`, `Shipping Info Entered`, `Contact Info Entered`, `Payment Info Entered` and `Order Completed` `track()` events. `products.$.` indicates an array named 'products'. Note: If a property is missing it is due to Shopify not providing the value, or due to a `null` or empty value being provided.  

Reviewer: ```suggestion
## Checkout and purchase Track event properties

The properties below are included with `Checkout Started`, `Address Info Entered`, `Shipping Info Entered`, `Contact Info Entered`, `Payment Info Entered` and `Order Completed` Track events. `products.$.` indicates an array named 'products'. 

> info "Missing properties?"
> Your Segment events may be missing one of the following properties if Shopify provides `null` or an empty value for a property.

```

| `currency`                              | The three-letter code that represents the currency, for example 'USD'                          | String        |
| `order_id`                              | The ID for the order                                                                           | String        |
| `checkout_id`                           | A unique identifier for a particular checkout                                                  | String        |
| `shipping`                              | Total shipping price                                                                           | Number        |
| `discount`                              | Total monetary value allocated as a discount                                                   | Number        |
| `coupon`                                | Comma delimited list of customer-facing discount names / codes                                 | String        |
| `payment_method`                        | Comma delimited list of payment providers used for the transaction                             | String        |

## Sending user profile details to Segment

User profile details can be sent to Segment in 3 ways: 

> 1. As `traits` in `identify()` events.
> 2. As `properties` in `track()` events. 
> 3. As `traits` in `track()` events, included in the `context.traits` object. 

Reviewer: ```suggestion
1. As `traits` in Identify events.
2. As `properties` in Track events. 
3. As `traits` in Track events, included in the `context.traits` object.
```

User profile details are only sent to Segment when the following track events are triggered by Shopify: 
`Checkout Started`, `Address Info Entered`, `Shipping Info Entered`, `Contact Info Entered`, `Payment Info Entered` and `Order Completed`.

> You can control how user profile details are sent to Segment via the 'Manage user profile details' section in the Shopify App User Interface.  

Reviewer: ```suggestion
You can control how user profile details are sent to Segment on the 'Manage user profile details' section of the Shopify App User Interface.
```

![Screenshot of the Manage user profile details settings section.](images/manage_user_profile_details.png)

> The following user profile details will be sent to Segment as they become available in the user checkout journey.  

Reviewer: ```suggestion
The following user profile details are sent to Segment as they become available in the user checkout journey.  
```

| `billing_address.postal_code`      | Billing address post code / zip code                                |
| `billing_address.province_code`    | Billing address province or state code                              |
| `billing_address.country`          | Billing address country                                             |
| `shipping_address.address1`        | Shipping address first line                                         |
| `shipping_address.address2`        | Shipping address second line                                        |
| `shipping_address.city`            | Shipping address city                                               |
| `shipping_address.country`         | Shipping address country                                            |
| `shipping_address.postal_code`     | Shipping address post code / zip code                               |
| `shipping_address.province_code`   | Shipping address province or state code                             |
| `shipping_address.country`         | Shipping address country                                            |

## FAQs

> ** Under reporting of events **
> This Source uses client side Javascript to send data to Segment. As with any client side analytics tracking it is possible that some ad-blockers will prevent data being sent to Segment. This can lead to an under reporting of events to Segment.   

Reviewer: ```suggestion
### Why don't I see all of my events?
This Source uses client-side Javascript to send data to Segment. As with any client-side analytics tracking, some ad-blockers prevent data from being sent to Segment. This can lead to an under reporting of events to Segment. For more information about data loss associated with ad blockers, see the [Ad Blocking](/docs/connections/sources/catalog/libraries/website/javascript/#ad-blocking) documentation. 
```

> ** When do identify() events trigger? **
> `identify()` events will only trigger when `Checkout Started`, `Address Info Entered`, `Shipping Info Entered`, `Contact Info Entered`, `Payment Info Entered` or `Order Completed` `track()` are triggered. Ensure that the 'Send identify() calls' checkbox is enabled. 

Reviewer: ```suggestion
### When do Identify events trigger?
To send Identify events, you must select the 'Send identify() calls' checkbox. If this setting is selected, Identify events are triggered when `Checkout Started`, `Address Info Entered`, `Shipping Info Entered`, `Contact Info Entered`, `Payment Info Entered` or `Order Completed` Track calls are triggered.
```

> ** identify() events sometimes don't trigger when expected **
> To save on API call volume Segment will only trigger an `identify()` event when user profile details have changed. If no change is detected then the `identify()` call will not trigger. 

Reviewer: ```suggestion
### Why aren't my Identify events triggering when expected?
To save on API call volume, Segment only triggers an Identify event when user profile details have changed. If no change is detected, then the Identify call is not triggered.  
```

> ** Why are some events duplicated and triggered multiple times? **

Reviewer: ```suggestion
### Why are some events duplicated and triggered multiple times?
```

Shopify sometimes erroneously triggers duplicate `Address Info Entered`, `Shipping Info Entered`, `Contact Info Entered`, `Payment Info Entered` events in short succession. This is a known bug with Shopify. Segment deduplicates the majority of these duplicate events but is unable to filter all of them out. 

> ** Why do some Address/Shipping/Contact/Payment Info Entered events contain product arrays? **
> Shopify sometimes includes product array details in the [Standard Events](https://shopify.dev/docs/api/web-pixels-api/standard-events){:target="\_blank"} which Segment uses to populate the `Address Info Entered`, `Shipping Info Entered`, `Contact Info Entered`, `Payment Info Entered` events. A decision was made to include these product details in the Segment events for completeness.  

Reviewer: ```suggestion
### Why do some Address/Shipping/Contact/Payment Info Entered events contain product arrays?
Shopify sometimes includes product array details in [Standard Events](https://shopify.dev/docs/api/web-pixels-api/standard-events){:target="_blank"} which Segment uses to populate the `Address Info Entered`, `Shipping Info Entered`, `Contact Info Entered`, `Payment Info Entered` events. Segment includes these product details in Segment events.
```

> ** Why are some event properties missing even though this documentation indicates that they should be present? **
> Mapping code is in place to map the majority of valuable fields from Shopify's [Standard Events](https://shopify.dev/docs/api/web-pixels-api/standard-events){:target="\_blank"} to Segment events per this document. If properties are missing it is due to the values not being provided by the Shopify standard event.   

Reviewer: ```suggestion
### Why are some event properties missing even though the documentation indicates that they should be present?
Mapping code is in place to map the majority of valuable fields from Shopify's [Standard Events](https://shopify.dev/docs/api/web-pixels-api/standard-events){:target="_blank"} to Segment events. If properties are missing, it is due to the values not being provided by the Shopify event.   
```
